src/viam/sdk/robot/client.hpp

```diff
@@ -44,6 +41,58 @@ using viam::robot::v1::Status;
 /// `with_channel` require a user call to `close()`.
 class RobotClient {
    public:
+    struct discovery_query {
+        viam::robot::v1::DiscoveryQuery to_proto() const;
+        static discovery_query from_proto(const viam::robot::v1::DiscoveryQuery& proto);
```

Comment:
It occurs to me that the goal of this PR is to remove proto types from the `robot_client` APIs, but these structs are reintroducing a proto dependency. This isn't great! I wanna think about the best way to go about resolving this... perhaps it is best to introduce a `Robot` base class after all. We can define the structs in `Robot`, and then define their `to_proto`/`from_proto` methodology in `RobotClient`. 

Another possibility: I _believe_ the only users of the `to_proto`/`from_proto` methods here are the robot tests. If that is correct (which will need to be confirmed). That means possibly we could remove those tests, or have them include `robot/client.*cpp*` rather than `robot/client.hpp` in order to access methods not exposed in the `.hpp`. Honestly both these options seem pretty bad and probably worth just discarding (especially when the `Robot` base class option exists), but I'm curious if others have thoughts or other proposals.

Reply:
Yes, I mentioned this in an earlier round of review too. Ultimately, this review isn't currently achieving the stated goal of removing the proto types from the `Robot` API: it has just pushed them into the `to_proto` and `from_proto` methods on the nested types.

One thing you could do is just make all of them into free functions in an unnamed namespace inside `robot.cpp`:

```
// robot.cpp
...
namespace {
viam::robot::v1::DiscoveryQuery to_proto(const RobotClient::discovery_query& dq) { ... };
discovery_query from_proto(const viam::robot::v1::DiscoveryQuery& proto);

// Repeat for `discovery`, `frame_system_config`, etc.

}  // namespace
```

This will work very well for implementing `Robot` and wouldn't force the introduction of a base class for Robot.

However, it comes with one notable downside, which is that the conversion functions are no longer available to the tests.

That's frustrating, but maybe not the end of the world. After all, the tests are perfectly able to traffic in the proto types. So if you can arrange a suitable situation via mocks, then you can test indirectly: call the API with the vocabulary type, collect the proto object in the mock, and compare. Then do the reverse.

After all, the whole point of having a class like `Robot` is to encapsulate the proto stuff away and provide a better UX to consumers.

Reply:
So the move seems to be to make a base class `Robot` in a new `robot.cpp` file. As for testing, `test_robot` doesn't use the protos, but `test_types` does.

@acmorrow When you say "collect the proto object in the mock, and compare", do you mean having `test_types` include the `robot.cpp` file that has the base `Robot` class? I changed the mocks so that they return the custom defined types, but I can easily define them in `test_type`.

Reply:
No, let's stay away from `#include`ing `.cpp` files: it is a quick way to end up with multiply defined symbols.

Instead, my suggestion is to make all of the proto conversion functions private to the `robot.cpp` file. Tests will no longer be able to explicitly test them.

Instead, use the "mock pipeline" approach to connect the `RobotClient` to a mock. Then, inside the mock, you can validate that the vocabulary types work correctly end-to-end without ever using the proto types directly: that'll happen only in the `RobotClient` code. So the testing will still happen, but it will be indirect.

@stuqdog - Does that make sense to you?

Reply:
Yep, completely agree.

Reply:
Actually to clarify: I think the proto conversion functions can't be just private in `robot.cpp` if `Robot` is an abstract base class. Instead I think they should be private in `robot/client.cpp` where they are actually being used.

Reply:
Yep agreed, a project for a future ticket.